File: server/src/engine/BeliefState.ts
/**
 * BeliefState (IP-2010) — content-producing half: tasking + staleness decay.
 * computeOpponentView/applyDeception (the enforcement half) are added by IP-6010 to this same
 * file, per FS-106's own scope split (see that package's Documentation Updates note).
 */
import type {
  Asset,
  AssetId,
  BeliefPrecision,
  BeliefStateEntry,
  ChainRole,
  OpponentView,
  OrbitalRegimeLabel,
  PlayerId,
  PlayerState,
} from '@owchess/shared';

const PRECISION_ORDER: BeliefPrecision[] = ['find', 'fix', 'track', 'target'];
/** BL-0009: turns without refresh before an entry decays one precision level. */
const STALENESS_WINDOW = 5;

function precisionIndex(p: BeliefPrecision): number {
  return PRECISION_ORDER.indexOf(p);
}

/**
 * True if chainRoles includes at least one F2T2E sensing role (find/fix/track/target) — false
 * for a purely 'engage' (effector-only) asset, which has nothing to task with (VR-2010 F1: a
 * caller must check this before spending AP, since applyTasking itself only no-ops silently).
 */
export function hasSensorCapability(chainRoles: ChainRole[]): boolean {
  return capabilityCeiling(chainRoles) >= 0;
}

/** The highest F2T2E precision level a sensor's declared chainRoles can produce. */
function capabilityCeiling(chainRoles: ChainRole[]): number {
  let ceiling = -1;
  for (const role of chainRoles) {
    const idx = PRECISION_ORDER.indexOf(role as BeliefPrecision);
    if (idx > ceiling) ceiling = idx;
  }
  return ceiling;
}

export class BeliefState {
  /**
   * FR-2100/2200: advances the observer's belief about whichever opponent asset(s) actually sit
   * in targetRegime, one precision level, capped at sourceAsset's chainRoles ceiling. A tasking
   * action that finds nothing in the regime produces no belief entry.
   */
  applyTasking(
    observer: PlayerId,
    observerState: PlayerState,
    sourceAsset: Asset,
    targetRegime: OrbitalRegimeLabel,
    opponentTrueState: PlayerState,
    turnNumber: number
  ): void {
    const ceiling = capabilityCeiling(sourceAsset.chainRoles);
    if (ceiling < 0) return; // no sensor role at all — nothing this asset can task with

    const presentAssets = [opponentTrueState.king, ...opponentTrueState.assets].filter(
      (a) => !a.destroyed && a.trueRegime === targetRegime
    );

    for (const target of presentAssets) {
      const existing = observerState.beliefOfOpponent.get(target.assetId);
      const currentIdx = existing ? precisionIndex(existing.precision) : -1;
      const nextIdx = Math.min(currentIdx + 1, ceiling, PRECISION_ORDER.length - 1);
      const precision = PRECISION_ORDER[Math.max(nextIdx, 0)];
      const entry: BeliefStateEntry = {
        subject: target.assetId,
        precision,
        lastUpdatedTurn: turnNumber,
        sourceAssetId: sourceAsset.assetId,
        deceived: existing?.deceived ?? false,
        apparentRegime: precisionIndex(precision) >= precisionIndex('fix') ? targetRegime : null,
      };
      observerState.beliefOfOpponent.set(target.assetId, entry);
    }
  }

  /**
   * FR-2300 (BL-0009): entries unrefreshed for >= 5 turns downgrade one level; a stale 'find'
   * entry is removed entirely (reverts to fully unknown), not floored.
   */
  decayStaleEntries(observerState: PlayerState, currentTurn: number): void {
    for (const [key, entry] of observerState.beliefOfOpponent) {
      if (currentTurn - entry.lastUpdatedTurn < STALENESS_WINDOW) continue;
      const idx = precisionIndex(entry.precision);
      if (idx <= 0) {
        observerState.beliefOfOpponent.delete(key);
        continue;
      }
      const downgraded: BeliefStateEntry = {
        ...entry,
        precision: PRECISION_ORDER[idx - 1],
        lastUpdatedTurn: currentTurn, // reset the clock on the downgraded level itself
        apparentRegime: idx - 1 >= precisionIndex('fix') ? entry.apparentRegime : null,
      };
      observerState.beliefOfOpponent.set(key, downgraded);
    }
  }

  /**
   * FR-6100/NFR-3100 (IP-6010): the SOLE function permitted to construct client-bound opponent
   * data. Reads only observerState.beliefOfOpponent (what observer has earned) — never reads
   * trueOpponentState's assets/king/activeEffects directly, which is what makes an accidental
   * full-object leak structurally impossible rather than merely a missed filter.
   */
  computeOpponentView(
    observer: PlayerId,
    observerState: PlayerState,
    trueOpponentState: PlayerState,
    _turnNumber: number
  ): OpponentView {
    return {
      playerId: trueOpponentState.playerId,
      beliefEntries: Array.from(observerState.beliefOfOpponent.values()),
    };
  }

  /**
   * FR-4003 (Deceive path, GDS-04/07's structural distinction): corrupts observer's own belief
   * entry for subject — never touches subject's true Asset record. Distinct code path from
   * Destroy (which removes/flags the true Asset itself, EffectResolver's job, not BeliefState's).
   */
  applyDeception(
    observerState: PlayerState,
    subject: AssetId,
    falseRegime: OrbitalRegimeLabel,
    turnNumber: number,
    sourceAssetId: AssetId
  ): void {
    const existing = observerState.beliefOfOpponent.get(subject);
    const entry: BeliefStateEntry = {
      subject,
      precision: existing?.precision ?? 'fix',
      lastUpdatedTurn: turnNumber,
      sourceAssetId: existing?.sourceAssetId ?? sourceAssetId,
      deceived: true,
      apparentRegime: falseRegime,
    };
    observerState.beliefOfOpponent.set(subject, entry);
  }
}

File: python-mvp/engine/belief_state.py
"""
BeliefState — per-player derived belief-state of the opponent; the fog-of-war boundary.
Mirrors server/src/engine/BeliefState.ts (tasking + staleness decay + the enforcement half:
computeOpponentView/applyDeception).

Sprint-0 fix (Deceive direction, per CLAUDE.md's documented defect): the TypeScript engine called
applyDeception(effectorObserverState, target.assetId, falseRegime ?? target.trueRegime, ...) — the
attacker corrupting *its own* belief entry about the opponent's asset with that asset's *real*
regime (since the client never sent falseRegime). That's not deception, it's a no-op that
happens to re-confirm the truth. A five-D "deceive" effect is EW doctrine for planting false
sensor returns in the victim's picture, so the corrected direction is: the effector plants a
false position for *itself* in the *target's owner's* belief map — beliefState.applyDeception is
now called with the target's owner's PlayerState (the party being deceived), subject =
effector.asset_id (the asset whose apparent position is now false), and a false_regime the
attacking player chooses in the Engage form (defaulting to a regime different from the
effector's true one if the player leaves it unset). This is a Sprint-0 spec call, flagged for
06-feature-specification once the pipeline resumes, per the task brief.
"""
from __future__ import annotations

from .types import Asset, BeliefStateEntry, ChainRole, OrbitalRegimeLabel, PlayerState

PRECISION_ORDER = ["find", "fix", "track", "target"]
STALENESS_WINDOW = 5  # turns without refresh before an entry decays one precision level


def _precision_index(p: str) -> int:
    return PRECISION_ORDER.index(p)


def has_sensor_capability(chain_roles: list[ChainRole]) -> bool:
    return _capability_ceiling(chain_roles) >= 0


def _capability_ceiling(chain_roles: list[ChainRole]) -> int:
    ceiling = -1
    for role in chain_roles:
        if role in PRECISION_ORDER:
            ceiling = max(ceiling, PRECISION_ORDER.index(role))
    return ceiling


class BeliefState:
    def apply_tasking(
        self,
        observer_state: PlayerState,
        source_asset: Asset,
        target_regime: OrbitalRegimeLabel,
        opponent_true_state: PlayerState,
        turn_number: int,
    ) -> None:
        ceiling = _capability_ceiling(source_asset.chain_roles)
        if ceiling < 0:
            return

        present = [
            a
            for a in opponent_true_state.all_assets()
            if not a.destroyed and a.true_regime == target_regime
        ]
        for target in present:
            existing = observer_state.belief_of_opponent.get(target.asset_id)
            current_idx = _precision_index(existing.precision) if existing else -1
            next_idx = min(current_idx + 1, ceiling, len(PRECISION_ORDER) - 1)
            precision = PRECISION_ORDER[max(next_idx, 0)]
            entry = BeliefStateEntry(
                subject=target.asset_id,
                precision=precision,
                last_updated_turn=turn_number,
                source_asset_id=source_asset.asset_id,
                deceived=existing.deceived if existing else False,
                apparent_regime=target_regime if _precision_index(precision) >= _precision_index("fix") else None,
            )
            observer_state.belief_of_opponent[target.asset_id] = entry

    def decay_stale_entries(self, observer_state: PlayerState, current_turn: int) -> None:
        for key in list(observer_state.belief_of_opponent.keys()):
            entry = observer_state.belief_of_opponent[key]
            if current_turn - entry.last_updated_turn < STALENESS_WINDOW:
                continue
            idx = _precision_index(entry.precision)
            if idx <= 0:
                del observer_state.belief_of_opponent[key]
                continue
            entry.precision = PRECISION_ORDER[idx - 1]
            entry.last_updated_turn = current_turn
            if idx - 1 < _precision_index("fix"):
                entry.apparent_regime = None

    def compute_opponent_view(self, observer_state: PlayerState) -> list[BeliefStateEntry]:
        return list(observer_state.belief_of_opponent.values())

    def apply_deception(
        self,
        victim_state: PlayerState,
        subject_asset_id: str,
        false_regime: OrbitalRegimeLabel,
        turn_number: int,
        source_asset_id: str,
    ) -> None:
        existing = victim_state.belief_of_opponent.get(subject_asset_id)
        entry = BeliefStateEntry(
            subject=subject_asset_id,
            precision=existing.precision if existing else "fix",
            last_updated_turn=turn_number,
            source_asset_id=existing.source_asset_id if existing else source_asset_id,
            deceived=True,
            apparent_regime=false_regime,
        )
        victim_state.belief_of_opponent[subject_asset_id] = entry
